Update MovingWindow to add resampler

daniel-zullo-frequenz opened the pull request:
Fixes #269

daniel-zullo-frequenz committed:
Fix duplicated word in documentation

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update MovingWindow size parameter type to timedelta

Modify the MovingWindow class to accept the size parameter
as a timedelta object instead of an integer. This change
allows the size parameter to represent the time span of
the moving window over which samples will be stored,
making it more intuitive for users and consistent among
other parameters in the MovingWindow class.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
This is a draft which only updates the size parameter type to timedelta of the MovingWindow at the moment.
I'm working out the changes related to the resampler and I'll report here once it is done (hopefully today EOD)
fyi @matthias-wende-frequenz

daniel-zullo-frequenz committed:
Remove attribute from MovingWindow

There is not need at the moment to keep the
state of the flag to prevent deep copying of
the adjacent buffer in MovingWindow.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

mathias-baumann-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -109,9 +118,19 @@ def __init__(
+            sampling_period <= size
+        ), "The sampling period should be equal to or lower than the window size."
+
+        # Add one more sample if sampling_period does not fit perfectly into the window size
```
Could you describe the motivation instead of the function? :)

daniel-zullo-frequenz replied:
of course

daniel-zullo-frequenz replied:
already done

daniel-zullo-frequenz committed:
Add resampler to the MovingWindow

Add a resampler in the MovingWindow to control
the granularity of the samples to be stored in
the underlying buffer.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Re-raise exception if MovingWindow task is cancelled

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -109,33 +121,84 @@ def __init__(
+        async def sink_buffer(sample: Sample) -> None:
+            self._buffer.update(sample)
+
+        avg_channel = Broadcast[Sample]("average")
```
This should Unicast instead

shsms replied:
We don't have a Unicast implementation.  Broadcast is a superset.

daniel-zullo-frequenz replied:
oh I see. I wrongly understood it. Then this will be kept as it is, thanks for letting me know!

leandro-lucarella-frequenz replied:
There was an attempt to make a unicast channel but it was slower than a broadcast channel. At some point we could add a unicast channel using a broadcast channel under the hood to provide an interface that only has one sender and one receiver, just for the sake of having a restricted interface.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -217,7 +217,7 @@ def __getitem__(self, key: SupportsIndex | datetime | slice) -> float | ArrayLik
             # see also (https://peps.python.org/pep-0696)
             if isinstance(key.start, datetime) and isinstance(key.stop, datetime):
-                return self._buffer.window(key.start, key.stop, self._copy_buffer)
+                return self._buffer.window(key.start, key.stop, force_copy=False)
```
the `force_copy` parameter can be removed from that function, since `False` is the default.

IIRC I left it in, because I wanted to expose this to the user but didn't had a plan at this time but I think it's fine to remove it until it's really used.

daniel-zullo-frequenz replied:
done, updated as suggested

matthias-wende-frequenz reviewed (commented):
This is looking quite good already. I'll have another look once the commit title doesn't say dirty any more :).

daniel-zullo-frequenz commented:
> This is looking quite good already. I'll have another look once the commit title doesn't say dirty any more :)

Cool, good to know that! Then it will 'beautify' the current changes and add units tests. I'll let you know once it's ready to review

shsms commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -110,6 +112,7 @@ def __init__(
             resampled_data_recv: A receiver that delivers samples with a
                 given sampling period.
             sampling_period: The sampling period.
+            resampling_period: The resampling period.
```
I'm wondering if we should rename these to `input_sampling_period` and `output_sampling_period` or at least improve the documentation.

leandro-lucarella-frequenz replied:
Yeah, it confused me too. In the resampler I call the input sampling period "source sampling period", but I guess it's different, because the resampler doesn't control this, while this class controls both, right?

daniel-zullo-frequenz replied:
correct, this class controls both.
Also keep in mind the `resampling_period` or `output_sampling_period` is optional.
Anyway it's fine with me renaming both input parameters and updating the documentation too.

daniel-zullo-frequenz replied:
I've initially done the renaming but then I thought it will be more suitable to pass the `ResamplerConfig` as it is more flexible to set different functionality and it also provides default parameters which fit well the current use case of the MovingWindow

shsms replied:
I think the `ResamplerConfig` has many features related to resampling of raw irregular streams,  that are not relevant for the second-level resampling that we're doing here and have to be ignored.

I think that would make the interface confusing, and people would have to read docs elsewhere,  outside the `ResamplerConfig` docs, to know which fields are relevant.

It would be possible to be far more precise if we just create a new `MovingWindowConfig` type here instead.

daniel-zullo-frequenz replied:
The way I see it is the `ResamplerConfig` provides all the default parameters suitable for the `MovingWindow` given the current use-case, and I found it useful and helpful the fact that the user can provide another resampling function through the `ResamplerConfig` along with the resampling period if needed.
Anyway I'm happy to address it is the way you think it is the best as you have more experience with the users of the SDK. @leandro-lucarella-frequenz @matthias-wende-frequenz any thought on this matter?

shsms replied:
I was thinking this was something else.  Based on the clarifications from the meeting,  I'm ok with going forward with this.

shsms commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -125,20 +128,39 @@ def __init__(
+        if resampling_period is not None:
+            assert (
+                resampling_period >= sampling_period
+            ), "Resampling must be equal to or higher than sampling, only down-sampling supported."
```
Could you add this limitation to the class docstring as well?  And probably also mention that resampling is also possible.

leandro-lucarella-frequenz replied:
Also, why the limitation?

daniel-zullo-frequenz replied:
I understood there is not use cases for the `MovingWindow` to support up-sampling at the moment, I was wondering about it and asked in slack last week.
Our common/typical scenario involves combining 2 data streams from different sources, e.g. meters, where one stream has a higher sampling period. So implementing down-sampling helps to preserve the precision of the data.

So should I just update the docs? Or add support for up-sampling and update the docs?

shsms replied:
That makes sense,  it is fine to say for now that the best they can get is the same rate as the resampling actor.  I'd say we discuss up-sampling if someone asks for it.

daniel-zullo-frequenz replied:
I removed the limitation as in the current patch the user can pass the ResamplerConfig, but I'm fine to revert it if we don't want that.

shsms replied:
I think it would be a good idea to remove it for now, because we don't see that there would be any use-cases for this.  And look into it later when we come across an actual use-case for it.

daniel-zullo-frequenz replied:
rigth, this was already removed.

shsms commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -149,16 +171,34 @@ async def _run_impl(self) -> None:
+
+    def _configure_resampler(self) -> None:
+        """Configure the components needed to run the resampler."""
+        assert isinstance(self._resampler, Resampler)
```
I think the more idiomatic way to do this is: `assert self._resampler is not None`

daniel-zullo-frequenz replied:
done, updated as suggested

daniel-zullo-frequenz committed:
Fix typos in documentation for Resampler

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add timezone to datetime types

There were inconsistencies in the way the SDK set
datetime variables in the MovingWindow/RingBuffer without
tz/tzinfo and in the Resampler with tz/tzinfo.
This patch sets the timezone UTC for all datetime variables
set in the MovingWindow and RingBuffer to make them consistent
to the datetime variables set in the Resampler.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Test resampling in MovingWindow

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
@matthias-wende-frequenz this is now ready to review

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -49,7 +50,11 @@ class MovingWindow:
+        size=timedelta(minutes=5),
+        resampled_data_recv=resampled_data_recv,
+        sampling_period=timedelta(seconds=1)
+    )
```
This change seems to be in the wrong commit, it doesn't seem to be related to changing the period to `timedelta`.

daniel-zullo-frequenz replied:
yeah, good catch! It probably happened when I rebased to applied the latest changes. I'll amend it

daniel-zullo-frequenz replied:
Fixed

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -46,14 +47,24 @@ class MovingWindow:
         size=timedelta(minutes=5),
         resampled_data_recv=resampled_data_recv,
-        sampling_period=timedelta(seconds=1)
+        input_sampling_period=timedelta(seconds=1)
```
Maybe mention this change in name (breaking change) explicitly in the release notes?

daniel-zullo-frequenz replied:
done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -91,11 +102,12 @@ class MovingWindow:
         resampled_data_recv: Receiver[Sample],
-        sampling_period: timedelta,
+        input_sampling_period: timedelta,
+        resampler_config: ResamplerConfig | None = None,
```
Just a mental note I want to share. I was in doubt if creating the `ResamplerConfig`, as it added one more layer when creating a resampler (`Resampler(ResamplerConfig(...))` instead of just `Resampler(...)`) but now seeing it used, I think it was a huge win, it is very easy to use the resampler.

Without it it would be a nightmare to provide the full flexibility to configure it, having to repeat many very specific configuration options and forward them, and duplicate the documentation, or just give up and only take a resampling period, and lose all the resampler flexibility in the way.

I think we should apply this pattern in any class that can be created using some sort of non-trivial configuration.

daniel-zullo-frequenz replied:
Well, I had a similar thought when Sahas showed the code in #270 and found it very convenient considering flexibility. I believe it will scale well with futures uses cases

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -119,26 +132,43 @@ def __init__(
+                resampling_period <= size
+            ), "The resampling period should be equal to or lower than the window size."
+
+            if resampling_period != input_sampling_period:
+                self._resampler = Resampler(resampler_config)
+                sampling = resampling_period
```
I'm not sure if this is a good idea to skip the resampling in this case. What if the user wants to configure the resampler differently? Like use a different resampling function? If it passed a resampler config explicitly with the same period it might be for a reason.

I don't know, maybe it doesn't make sense to resample at all if the input sampling rate is the same as the requested resampling rate, but ignoring it silently makes some noise in my head, if I explicitly pass a config it should be used IMHO. Another option would be to raise an exception if they match.

daniel-zullo-frequenz replied:
I think it makes sense to remove the check and I wouldn't add an assertion given that user might just want to use the same resampling period with a different resampling function as you also mentioned it

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -119,26 +132,43 @@ def __init__(
 
         # Sampling period might not fit perfectly into the window size.
-        num_samples = math.ceil(size / sampling_period)
+        num_samples = math.ceil(size.total_seconds() / sampling.total_seconds())
```
I know it is very simple code, but I've seen this before. Maybe move it to a `_get_num_samples()` utility method to make sure the way is calculated is consistent everywhere we calculate it?

daniel-zullo-frequenz replied:
This is a special case and it is only used to set the size of the underlying ring buffer when it is instantiated. A user can get the size of the moving window by calling len(moving_window) as `__len__` was implemented (recently fixed in #274)

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -149,16 +179,35 @@ async def _run_impl(self) -> None:
+            if sample.value is not None:
+                self._buffer.update(sample)
+
+        avg_channel = Broadcast[Sample]("average")
+        self._avg_sender = avg_channel.new_sender()
+        self._resampler.add_timeseries("avg", avg_channel.new_receiver(), sink_buffer)
+        self._resampler_task = asyncio.create_task(self._resampler.resample())
```
Looking at this, I'm starting to feel that we should have a resampler that don't work on channels at all, it makes things more complicated (an inefficient) to go through a channel here, where you could have just call `resampler.add(sample)` when handling the data:

```py
            async for sample in self._resampled_data_recv:
                log.debug("Received new sample: %s", sample)
                if self._resampler:
                    self._reampler.add(sample)
                else:
                    self._buffer.update(sample)
```

There is already a `_ResampleHelper` to handle the resampling of only one metric, but it doesn't run its own timer, so we would need to have something that does handle its own timer and ideally can resample multiple metrics.

If you agree this is a good idea, I will turn this into an issue.

Big parenthesis: with historical data, we'll need to build windows without a timer, just assuming data comes ordered and looking at the sample timestamp to trigger the building of a new resampling window, so maybe we don't even need the timer after all.

daniel-zullo-frequenz replied:
Actually I was looking for a way to avoid using the channel for this particular use case when I started the implementation. I agree on the idea in a separate issue to move faster with this current PR, and then we can get rid of channel here once the issue is implemented. @matthias-wende-frequenz @sahas-subramanian-frequenz do you agree?

matthias-wende-frequenz replied:
Agreed!

leandro-lucarella-frequenz replied:
Actually the resampler already support this, because it doesn´t depend on channels but on async iterators. So it is a bit convoluted, but you could have an [async generator object](https://peps.python.org/pep-0525/#asynchronous-generator-object) to send samples to the resampler without creating a channel.

It will still be async, so it's not as efficient as it could be, and also it is convoluted to use, so I think we should still go for a simpler solution.

* https://github.com/frequenz-floss/frequenz-sdk-python/issues/303

daniel-zullo-frequenz replied:
Good to know that! Then I think it is fine to keep the resampler here in the current state and get rid of the channel once #303 is implemented

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -46,14 +47,24 @@ class MovingWindow:
     documentation.
 
+    Resampling might be required when working with streams of samples from
+    different sources at different sampling rates or periods, and it can be
```
This is not quite accurate. At this point in time streams are already combined. Here it's mostly useful to reduce the amount of data.

daniel-zullo-frequenz replied:
You're right, I'll amend it and probably need to amend the release notes

daniel-zullo-frequenz replied:
Amended

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -119,26 +132,43 @@ def __init__(
+
+        sampling = input_sampling_period
+        self._resampler: Resampler | None = None
+        self._avg_sender: Sender[Sample] | None = None
```
This `avg_sender` got me really confused. This is just the channel where to send samples to get resampled by the resampler, right? If so, I would name it `resampling_source` or `resampling_sender` to make it more clear it only exists for resampling.

daniel-zullo-frequenz replied:
You're right, `resampling_sender` sounds good to me

daniel-zullo-frequenz replied:
Actually it seems it might be better to call it `resampler_sender` to be consistent with the other names e.g. resampler_config, resampler_task, etc

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -125,7 +126,7 @@ def __init__(  # pylint: disable=too-many-arguments
             window_alignment: A datetime object that defines a point in time to which
                 the window is aligned to modulo window size.
-                (default is midnight 01.01.01)
+                (default is midnight 01.01.01 UTC)
```
I know this is not introduced by this PR, but since you are touching it, I would update this to use ISO format for dates (0001-01-01T00:00:00+00:00), is not that it is super clear, but using this German short format is **very** confusing.

daniel-zullo-frequenz replied:
will update it as suggested.

matthias-wende-frequenz replied:
My Bad! Sorry about that ;).

daniel-zullo-frequenz replied:
I fixed it in a separate commit

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_ringbuffer.py:
```diff
@@ -76,8 +76,8 @@ def __init__(
         self._gaps: list[Gap] = []
-        self._datetime_newest: datetime = datetime.min
-        self._datetime_oldest: datetime = datetime.max
+        self._datetime_newest: datetime = datetime.min.replace(tzinfo=timezone.utc)
+        self._datetime_oldest: datetime = datetime.max.replace(tzinfo=timezone.utc)
```
Here `datetime.max` (and later `datetime.min` are used just as markers, right? I would define them as class constants for 2 reasons: We can change the way we represent the markers, and we don't need to do the `.replace()` every time we want to use them:

```py
class OrderedRingBuffer:
    _DATETIME_MIN = datetime.min.replace(tzinfo=timezone.utc)
    _DATETIME_MAX = datetime.max.replace(tzinfo=timezone.utc)
    
    def __init__(self, ...):
        ...
        self._datetime_newest: datetime = self._DATETIME_MIN
        self._datetime_oldest: datetime = self._DATETIME_MAX
```

In the future I would replace that with a sentinel as IMHO it is even more clear that it is used as a marker. With the above we could even use `is not` to check if it is the marker instead of `!=`. But in this case is not as straightforward because if we really hit the min/max datetime in the value of `self._datetime_oldest` or `self._datetime_newest` then we are screwed :D

daniel-zullo-frequenz replied:
Yes, they are markers. I'll change it as suggested in a separate commit.

matthias-wende-frequenz reviewed (dismissed):
One small comment about the usage of the resampler in this case. Otherwise it's looking good!

leandro-lucarella-frequenz reviewed (commented):
A few comments, but in general LGTM! :tada:

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -39,19 +41,34 @@ class MovingWindow:
+
+    If resampling is not required, the resampler config parameter can be
+    set to None or the resampling period can be set equal to the input
+    sampling period in which case the MovingWindow will not perform any
```
Also I might need to amend this paragraph as `or the resampling period can be set equal to the input sampling period` might no longer apply

daniel-zullo-frequenz committed:
Amend examples in MovingWindow documentation

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Define datetime min and max as constants

Change the way the markers are represented in RingBuffer
so that `replace()` is not called every time the markers
are assigned or compared to.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update documentation to use ISO format for dates

The ISO format is more precise and  might be less
confusing for the users.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
Updated bases on some suggestions from Luca and Matthias. There are still some on-going conversations

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_ringbuffer.py:
```diff
@@ -130,7 +133,10 @@ def update(self, sample: Sample) -> None:
-        if timestamp < self._datetime_oldest and self._datetime_oldest != datetime.max:
+        if (
+            timestamp < self._datetime_oldest
+            and self._datetime_oldest != datetime.max.replace(tzinfo=timezone.utc)
```
```suggestion
            and self._datetime_oldest != self._DATETIME_MAX
```
?

daniel-zullo-frequenz replied:
oops I missed that one

daniel-zullo-frequenz replied:
updated

leandro-lucarella-frequenz reviewed (commented):
Other than this oversight, LGTM!

daniel-zullo-frequenz commented:
Rebased onto latest v0.x.x branch

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -46,6 +47,13 @@ class MovingWindow:
     documentation.
 
+    Resampling might be required to reduce the number of samples to store
+    without losing precision, and it can be set by specifying the resampler
```
I'm not sure what you want to say without loosing precision. We are loosing precision for when we downsample but that's acceptable in many cases.

daniel-zullo-frequenz replied:
I meant that we can reduce the number of samples without losing *much* precision through resampling. I'd then remove `without losing precision` as it is confusing and it isn't generic in the sense it only applies for downsampling where it is somehow implicit an acceptable loss of precision

matthias-wende-frequenz reviewed (dismissed):
just one last comment :)

daniel-zullo-frequenz commented:
Updated considering the latest comment from Matthias

daniel-zullo-frequenz committed:
Increase sampling rate for a flaky test

The test fails when it is run in the CI, increasing
the sampling won't solve the problem, just it will
fail less ofter. There are already two different open
issues to address and fix these flaky tests. See the
links below:

https://github.com/frequenz-floss/frequenz-sdk-python/issues/249
https://github.com/frequenz-floss/frequenz-sdk-python/issues/70

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>
